planner/core/rule_join_elimination.go

```diff
@@ -173,38 +183,60 @@ func (o *outerJoinEliminator) isDuplicateAgnosticAgg(p LogicalPlan) (_ bool, col
 			return false, nil
 		}
 		for _, expr := range aggDesc.Args {
-			if col, ok := expr.(*expression.Column); ok {
-				cols = append(cols, col)
-			} else {
-				return false, nil
-			}
+			// ExtractColumns will trans the expr to cols.
+			cols = append(cols, expression.ExtractColumns(expr)...)
 		}
 	}
 	return true, cols
 }
 
-func (o *outerJoinEliminator) doOptimize(p LogicalPlan, aggCols []*expression.Column, parentSchema *expression.Schema) (LogicalPlan, error) {
+func (o *outerJoinEliminator) doOptimize(p LogicalPlan, aggCols []*expression.Column, parentCols []*expression.Column) (LogicalPlan, error) {
+	var err error
+	var isEliminated bool
+	for join, isJoin := p.(*LogicalJoin); isJoin; join, isJoin = p.(*LogicalJoin) {
+		p, isEliminated, err = o.tryToEliminateOuterJoin(join, aggCols, parentCols)
+		if err != nil {
+			return p, err
+		}
+		if !isEliminated {
+			break
+		}
+		aggCols = aggCols[:0]
+	}
+
+	var colsInSchema []*expression.Column
+	switch p.(type) {
+	case *LogicalProjection:
+		for _, expr := range p.(*LogicalProjection).Exprs {
+			colsInSchema = append(colsInSchema, expression.ExtractColumns(expr)...)
+		}
+	case *LogicalSelection:
+		for _, expr := range p.(*LogicalSelection).Conditions {
+			colsInSchema = append(colsInSchema, expression.ExtractColumns(expr)...)
+		}
+	default:
+		colsInSchema = p.Schema().Columns
+	}
+
 	// check the duplicate agnostic aggregate functions
 	if ok, newCols := o.isDuplicateAgnosticAgg(p); ok {
 		aggCols = newCols
+	} else if len(aggCols) > 0 {
+		aggCols = append(aggCols, colsInSchema...)
```

Comment:
This branch is to fix https://github.com/pingcap/tidb/issues/9536 ?
But i think the reason is that you don't check `isParentColsAllFromOuterTable` if you find that `isAggColsAllFromOuterTable` is true.

Reply:
`isParentColsAllFromOuterTable` and `isAggColsAllFromOuterTable ` are defferent eliminating rule. But your suggestion may also work, I think.

Reply:
Append column to `aggCols` is strange, i think this breaks what `aggCols` mean?